polkadot/runtime/parachains/src/inclusion/mod.rs

```diff
@@ -1028,67 +1016,81 @@ impl<T: Config> Pallet<T> {
-
-			if let (Some(pending), Some(commitments)) = (pending, commitments) {
-				// defensive: this should always be true.
-				let candidate = CandidateReceipt {
-					descriptor: pending.descriptor,
-					commitments_hash: commitments.hash(),
-				};
+		let timed_out: Vec<_> =
+			Self::free_cores(|candidate| pred(candidate.backed_in_number).timed_out, None)
+				.collect();
+		let mut timed_out_cores = Vec::with_capacity(timed_out.len());
+		for candidate in timed_out.iter() {
+			timed_out_cores.push(candidate.core);
+
+			let receipt = CandidateReceipt {
+				descriptor: candidate.descriptor.clone(),
+				commitments_hash: candidate.commitments.hash(),
+			};
 
-				Self::deposit_event(Event::<T>::CandidateTimedOut(
-					candidate,
-					commitments.head_data,
-					pending.core,
-				));
-			}
+			Self::deposit_event(Event::<T>::CandidateTimedOut(
+				receipt,
+				candidate.commitments.head_data.clone(),
+				candidate.core,
+			));
 		}
 
-		cleaned_up_cores
+		timed_out_cores
 	}
 
-	/// Cleans up all paras pending availability that are in the given list of disputed candidates.
+	/// Cleans up all cores pending availability occupied by one of the disputed candidates or which
+	/// are descendants of a disputed candidate.
 	///
-	/// Returns a vector of cleaned-up core IDs.
-	pub(crate) fn collect_disputed(disputed: &BTreeSet<CandidateHash>) -> Vec<CoreIndex> {
-		let mut cleaned_up_ids = Vec::new();
-		let mut cleaned_up_cores = Vec::new();
-
-		for (para_id, pending_record) in <PendingAvailability<T>>::iter() {
-			if disputed.contains(&pending_record.hash) {
-				cleaned_up_ids.push(para_id);
-				cleaned_up_cores.push(pending_record.core);
+	/// Returns a vector of cleaned-up core IDs, along with the evicted candidate hashes.
+	pub(crate) fn collect_disputed(
+		disputed: &BTreeSet<CandidateHash>,
+	) -> impl Iterator<Item = (CoreIndex, CandidateHash)> + '_ {
+		Self::free_cores(|candidate| disputed.contains(&candidate.hash), Some(disputed.len()))
+			.map(|candidate| (candidate.core, candidate.hash))
+	}
+
+	fn free_cores<P: Fn(&CandidatePendingAvailability<T::Hash, BlockNumberFor<T>>) -> bool>(
+		pred: P,
+		capacity_hint: Option<usize>,
+	) -> impl Iterator<Item = CandidatePendingAvailability<T::Hash, BlockNumberFor<T>>> {
+		let mut cleaned_up_cores =
+			if let Some(capacity) = capacity_hint { Vec::with_capacity(capacity) } else { vec![] };
+
+		for (para_id, pending_candidates) in <PendingAvailability<T>>::iter() {
+			// We assume that pending candidates are stored in dependency order. So we need to store
+			// the earliest dropped candidate. All others that follow will get freed as well.
+			let mut earliest_dropped_idx = None;
+			for (index, candidate) in pending_candidates.iter().enumerate() {
+				if pred(candidate) {
+					earliest_dropped_idx = Some(index);
+					// Since we're looping the candidates in dependency order, we've found the
+					// earliest disputed index for this paraid.
```

Comment:
Apparently you are assuming here to only be called for disputed candidates. That's not true, you also call this function for timed out.

In particular I would even reconsider the name for this function. What this function does is forcefully freeing cores - free cores that have _not_ become available. Which allows you to do the drain below, that would make little sense if we are freeing here based on availability. The name should reflect this, some documentation would also be good.

Maybe `cancel_cores` or `free_failed_cores`?

Reply:
I can document/comment this better. "earliest disputed index" was just an outdated comment

Reply:
done and renamed to `free_failed_cores `, have a look